3 course/OperatingSystem/http-proxy/http_proxy.c

```diff
@@ -0,0 +1,177 @@
+    struct sockaddr_in proxy_addr;
+    int proxy_socket;
+
+    proxy_socket = socket(AF_INET, SOCK_STREAM, 0);
+    if (proxy_socket < 0) {
+        perror("Ошибка при создании прокси-сокета");
+        exit(EXIT_FAILURE);
+    }
+
+    //настройка сокета для повторного использования порта
+    int opt = 1;
+    if (setsockopt(proxy_socket, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt))) {
+        perror("setsockopt");
+        exit(EXIT_FAILURE);
+    }
+
+    //настраиваем адрес и порт прокси-сокета
+    proxy_addr.sin_family = AF_INET;
+    proxy_addr.sin_addr.s_addr = INADDR_ANY;
+    proxy_addr.sin_port = htons(8080);
+
+    //привязываем прокси-сокет к адресу и порту
+    if (bind(proxy_socket, (struct sockaddr*)&proxy_addr, sizeof(proxy_addr)) < 0) {
+        perror("Ошибка при привязке прокси-сокета");
+        exit(EXIT_FAILURE);
+    }
+
+    //слушаем клиентиков
+    if (listen(proxy_socket, 5) < 0) {
+        perror("Ошибка при ожидании соединений");
+        exit(EXIT_FAILURE);
+    }
+
+    printf("[*] Прокси слушает на порту 80\n");
+
+    return proxy_socket;
+}
+
+void get_host(char* buffer, char* host) {
+    char buffer_copy[MAX_BUFFER_SIZE];
+    strcpy(buffer_copy, buffer);
+    char *first_line = strtok(buffer_copy, "\n");
+    char *url = strtok(first_line + 4, " ");
+    
+    int port;
+
+    sscanf(url, "http://%[^:/]:%d", host, &port);
+}
+
+int connect_and_get_server_socket(char *host) {
+    struct addrinfo hints, *result, *p;
+
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_family = AF_UNSPEC; 
+    hints.ai_socktype = SOCK_STREAM;
+
+    char* port = "80";
+    if (getaddrinfo(host, port, &hints, &result) != 0) {
+        perror("Ошибка при разрешении DNS");
+        exit(EXIT_FAILURE);
+    }
+
+    int server_socket;
+    for (p = result; p != NULL; p = p->ai_next) {
+        server_socket = socket(p->ai_family, p->ai_socktype, p->ai_protocol);
+        if (server_socket == -1) {
+            continue; // Ошибка при создании сокета, пробуем следующий адрес
+        }
+
+        if (connect(server_socket, p->ai_addr, p->ai_addrlen) != -1) {
+            break;
+        } else {
+            perror("Ошибка при установлении соединения с удаленным сервером");
+            close(server_socket);
+            return -1;
+        }
+    }    
+
+    freeaddrinfo(result);
+    return server_socket;
+}
+
+void handle_client(int client_socket) {
+    char buffer[MAX_BUFFER_SIZE];
+    ssize_t bytes_received;
+
+    // получаем запрос от клиента
+    bytes_received = recv(client_socket, buffer, sizeof(buffer), 0);
+    if (bytes_received < 0) {
+        perror("Ошибка при чтении из клиентского сокета");
+        close(client_socket);
+        return;
+    }
+
+    printf("RECV: %s\n", buffer);
+
+    char host[MAX_BUFFER_SIZE];
+    get_host(buffer, host);
+
+    int server_socket = connect_and_get_server_socket(host);
+
+    if (send(server_socket, buffer, bytes_received, 0) < 0) {
```

Comment:
Please handle signals correctly ,e.g. I'll stop your program many times  and then continue execution  but your connections will get dropped here

Reply:
?